### PR TITLE
ARROW-1491: [C++] Add casting from strings to numbers and booleans

### PR DESCRIPTION
**Pull request by pitrou**
The implementation for numbers uses C++ `istringstream`. This makes casting a bit lenient (it will probably accept whitespace).

This is a rewrite of https://github.com/apache/arrow/pull/1387

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/cast.cc [hunk 1]
```diff
@@ -727,6 +729,178 @@ struct CastFunctor<T, DictionaryType,
...
+  void operator()(FunctionContext* ctx, const CastOptions& options,
+                  const ArrayData& input, ArrayData* output) {
+    using out_type = typename O::c_type;
+
+    StringArray input_array(input.Copy());
+    auto out_data = GetMutableValues<out_type>(output, 1);
+    errno = 0;
+    // Instantiate the stringstream outside of the loop
+    std::istringstream ibuf;
+    ibuf.imbue(std::locale::classic());
```
Comment (cpcloud) [inline cpp/src/arrow/compute/kernels/cast.cc]: I sometimes wonder what the meeting was like that decided names like this.

Comment (pitrou) [reply cpp/src/arrow/compute/kernels/cast.cc]: I had the same thought :-)

File: cpp/src/arrow/compute/kernels/cast.cc [hunk 2]
```diff
@@ -727,6 +729,178 @@ struct CastFunctor<T, DictionaryType,
...
+      *out = true;
+      return true;
+    }
+    return false;
+  }
+  if (length == 4) {
+    // "true"?
+    *out = true;
+    return ((s[0] == 't' || s[0] == 'T') && (s[1] == 'r' || s[1] == 'R') &&
+            (s[2] == 'u' || s[2] == 'U') && (s[3] == 'e' || s[3] == 'E'));
```
Comment (cpcloud) [inline cpp/src/arrow/compute/kernels/cast.cc]: Can we make this a little more readable by use `tolower`?

Comment (pitrou) [reply cpp/src/arrow/compute/kernels/cast.cc]: `tolower` is locale-dependent.

File: cpp/src/arrow/compute/kernels/cast.cc [hunk 3]
```diff
@@ -727,6 +729,178 @@ struct CastFunctor<T, DictionaryType,
...
+    *out = static_cast<T>(res);  // may downcast
+    if (res > max_value) {
+      return false;
+    }
+  }
+  return !ibuf.fail() && ibuf.eof();
+}
+
+template <typename O>
+struct CastFunctor<O, StringType, typename std::enable_if<is_number<O>::value>::type> {
```
Comment (wesm) [inline cpp/src/arrow/compute/kernels/cast.cc]: You could use `enable_if_number<O>` here from arrow/type_traits.h

**Review by wesm — COMMENTED** [review]
Thanks for this. Made a small comment about reusing type traits but otherwise OK by me

File: cpp/src/arrow/compute/kernels/cast.cc (hunk 3 above)
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/cast.cc]: Done.

**Comment by pitrou** [comment]
Some weird failure occurred on the AppVeyor build that didn't occur on my AppVeyor account:
https://ci.appveyor.com/project/pitrou/arrow/build/1.0.570

**Comment by pitrou** [comment]
I want to make sure the AppVeyor failure was sporadic so I've triggered another build by pushing again.

**Comment by wesm** [comment]
rebased

**Review by wesm — APPROVED** [review]
+1

**Pull request closed without merging**